File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/operation/meta/RecordAttsService.java
package ru.citeck.ecos.records3.record.operation.meta;

import ru.citeck.ecos.records3.RecordAtts;
import ru.citeck.ecos.records3.source.common.AttMixin;

import java.util.Collection;
import java.util.List;
import java.util.Map;

public interface RecordAttsService {

    RecordAtts getAtts(Object value, Map<String, String> attributes);

    RecordAtts getAtts(Object value, Map<String, String> attributes, boolean rawAtts);

    RecordAtts getAtts(Object value, Collection<String> attributes);

    RecordAtts getAtts(Object value, Collection<String> attributes, boolean rawAtts);

    <T> T getAtts(Object value, Class<T> attsDto);

    <T> List<T> getAtts(List<?> values, Class<T> attsDto);

    List<RecordAtts> getAtts(List<?> values, Collection<String> attributes);

    List<RecordAtts> getAtts(List<?> values, Collection<String> attributes, boolean rawAtts);

    List<RecordAtts> getAtts(List<?> values, Map<String, String> attributes);

    List<RecordAtts> getAtts(List<?> values, Map<String, String> attributes, boolean rawAtts);

    List<RecordAtts> getAtts(List<?> values, Map<String, String> attributes, boolean rawAtts, List<AttMixin> mixins);
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/RecordAttsService.kt
package ru.citeck.ecos.records3.record.atts

import ru.citeck.ecos.records2.RecordRef
import ru.citeck.ecos.records3.record.atts.dto.RecordAtts
import ru.citeck.ecos.records3.record.atts.schema.SchemaAtt
import ru.citeck.ecos.records3.record.mixin.MixinContext

interface RecordAttsService {

    fun <T> doWithSchema(attributes: Map<String, *>, rawAtts: Boolean, action: (List<SchemaAtt>) -> T): T

    fun <T> doWithSchema(atts: List<SchemaAtt>, rawAtts: Boolean, action: (List<SchemaAtt>) -> T): T

    fun <T : Any> getAtts(value: Any?, attributes: Class<T>): T

    fun getAtts(value: Any?, attributes: Map<String, String>): RecordAtts

    fun getAtts(value: Any?, attributes: Map<String, String>, rawAtts: Boolean): RecordAtts

    fun getAtts(value: Any?, attributes: Collection<String>): RecordAtts

    fun getAtts(value: Any?, attributes: Collection<String>, rawAtts: Boolean): RecordAtts

    fun <T : Any> getAtts(values: List<*>, attributes: Class<T>): List<T>

    fun getAtts(values: List<*>, attributes: Collection<String>): List<RecordAtts>

    fun getAtts(values: List<*>, attributes: Collection<String>, rawAtts: Boolean): List<RecordAtts>

    fun getAtts(values: List<*>, attributes: Map<String, String>): List<RecordAtts>

    fun getAtts(values: List<*>, attributes: Map<String, String>, rawAtts: Boolean): List<RecordAtts>

    fun getAtts(values: List<*>, attributes: Map<String, String>, rawAtts: Boolean, mixins: MixinContext): List<RecordAtts>

    fun getAtts(values: List<*>, attributes: List<SchemaAtt>, rawAtts: Boolean, mixins: MixinContext): List<RecordAtts>

    fun getId(value: Any?, defaultRef: RecordRef): RecordRef

    fun getAtts(
        values: List<*>,
        attributes: List<SchemaAtt>,
        rawAtts: Boolean,
        mixins: MixinContext,
        recordRefs: List<RecordRef>
    ): List<RecordAtts>
}
